### ISSUE TITLE
Using @connectTo decorator throws error and crashes, although manually configuring bindings works fine

### ISSUE THREAD

**Issue by cmatzenbach** [aurelia/store#76]
**I'm submitting a bug report**

* **Library Version:**
1.1.4

**Please tell us about your environment:**
* **Operating System:**
Linux (arch)

* **Node Version:**
10.11.0

* **NPM Version:**
6.4.1

* **JSPM OR Webpack AND Version**
webpack 3.8.0

* **Browser:**
all

* **Language:**
TypeScript 3.1.3

**Current behavior:**
I set up aurelia-store exactly as described in the docs.  Everything worked fine, including setting up bindings to the store manually in a component ("Subscribing to the stream of states" section).  However, when I changed that file to use the @connectTo decorator as described in the next section, I get the following error: `TypeError: Cannot read property 'get' of undefined`.

I did some debugging, and found the bug in the file `node_modules/aurelia-store/dist/native-modules/decorator.js`.  The `import` statements on lines 12-14 are not working; the error that is being thrown is on line 20: `var store = Container.instance.get(Store);` - the problem is that `Container` is undefined (as well as all the other imported modules, I discovered).  What's even stranger, I can `import { Container } from 'aurelia-dependency-injection'` in my components just fine.  I looked in the package.json for the aurelia-store node_modules, and saw that it is using the latest version of aurelia-dependency-injector, so it doesn't seem like a versioning issue from what I have installed locally.

I thought the problem may be in having an old package, so I updated all my packages (not smart considering I use webpack, I know) and totally broke my application.  I've started upgrading just the aurelia packages one-by-one to see where the problem lies, and I can update here on that process if I have success in solving this.  I'll post my projects package.json file below:

**Expected/desired behavior:**
The @connectTo decorator works as in the documentation and doesn't throw an error.

package.json (some of the libraries in here are leftover and need to be cleaned out):
  "devDependencies": {
    "@types/aws4": "^1.5.1",
    "@types/cheerio": "^0.22.5",
    "@types/jest": "19.2.4",
    "@types/jquery": "^3.3.21",
    "@types/lodash": "^4.14.68",
    "@types/lodash-es": "^4.14.5",
    "@types/node": "^8.0.11",
    "@types/request-promise": "^4.1.39",
    "@types/underscore": "^1.8.0",
    "@types/webpack": "^3.0.4",
    "aurelia-loader-nodejs": "1.0.1",
    "aurelia-pal-nodejs": "1.0.0-beta.1.0.0",
    "aurelia-protractor-plugin": "1.0.3",
    "aurelia-testing": "1.0.0-beta.3.0.1",
    "aurelia-webpack-plugin": "2.0.0-rc.5",
    "autoprefixer": "7.1.1",
    "awesome-typescript-loader": "3.1.3",
    "aws-sdk": "^2.305.0",
    "babel-core": "^6.26.3",
    "babel-loader": "^7.1.4",
    "babel-plugin-uglify": "^1.0.2",
    "babel-preset-env": "^1.7.0",
    "babel-preset-es2015": "^6.24.1",
    "bootstrap-sass": "^3.3.7",
    "chai": "^4.1.2",
    "cheerio": "^1.0.0-rc.2",
    "compression-webpack-plugin": "^0.4.0",
    "copy-webpack-plugin": "4.0.1",
    "cross-env": "5.0.1",
    "css-loader": "0.28.4",
    "expose-loader": "0.7.3",
    "extract-text-webpack-plugin": "2.1.2",
    "file-loader": "0.11.2",
    "html-webpack-plugin": "2.29.0",
    "http-server": "0.10.0",
    "istanbul-instrumenter-loader": "2.0.0",
    "jasmine-core": "2.6.3",
    "jest": "20.0.4",
    "jest-cli": "20.0.4",
    "json-loader": "0.5.4",
    "karma": "1.7.0",
    "karma-babel-preprocessor": "^7.0.0",
    "karma-chrome-launcher": "2.1.1",
    "karma-coverage-istanbul-reporter": "1.3.0",
    "karma-jasmine": "1.1.0",
    "karma-mocha-reporter": "2.2.3",
    "karma-webpack": "2.0.3",
    "mocha": "^5.0.1",
    "mocha-webpack": "^1.0.1",
    "node-sass": "^4.9.3",
    "nps": "5.3.1",
    "nps-utils": "1.2.0",
    "postcss-loader": "2.0.5",
    "protractor": "^5.4.0",
    "purify-css": "^1.2.5",
    "purifycss-webpack": "^0.7.0",
    "sass-loader": "^6.0.6",
    "sinon": "^4.4.10",
    "style-loader": "0.18.2",
    "ts-jest": "20.0.6",
    "ts-node": "3.3.0",
    "typescript": "3.1.3",
    "url-loader": "0.5.8",
    "wait-on": "^2.1.0",
    "webpack": "^3.8.0",
    "webpack-dev-server": "2.7.1"
  },

  "dependencies": {
    "au-table": "^0.1.13",
    "aurelia-bootstrapper": "2.1.1",
    "aurelia-dialog": "^1.0.0-rc.2.0.0",
    "aurelia-fetch-client": "^1.1.3",
    "aurelia-framework": "^1.1.4",
    "aurelia-history-browser": "1.0.0",
    "aurelia-loader-webpack": "2.1.0",
    "aurelia-logging-console": "1.0.0",
    "aurelia-pal-browser": "1.2.1",
    "aurelia-plugins-cookies": "^2.5.1",
    "aurelia-polyfills": "^1.2.2",
    "aurelia-router": "1.3.0",
    "aurelia-store": "^1.2.0",
    "aurelia-templating": "1.4.2",
    "aurelia-templating-binding": "1.3.0",
    "aurelia-templating-resources": "1.4.0",
    "aurelia-templating-router": "1.1.0",
    "aurelia-validation": "^1.1.2",
    "axios": "^0.17.1",
    "babel-polyfill": "^6.26.0",
    "bluebird": "3.5.0",
    "bootstrap": "^4.0.0",
    "bootstrap-select": "1.13.0-beta",
    "corejs-typeahead": "^1.2.1",
    "cornerstone-core": "^2.2.4",
    "cornerstone-math": "^0.1.6",
    "cornerstone-tools": "^2.3.5",
    "cornerstone-web-image-loader": "^2.1.0",
    "dayjs": "^1.7.7",
    "dragula": "^3.7.2",
    "dropzone": "^5.4.0",
    "fast-xml-parser": "^3.11.2",
    "filepond": "^1.2.5",
    "filepond-plugin-file-validate-size": "^1.0.2",
    "filepond-plugin-file-validate-type": "^1.0.2",
    "filepond-plugin-image-crop": "^1.0.0",
    "filepond-plugin-image-exif-orientation": "^1.0.0",
    "filepond-plugin-image-preview": "^1.0.5",
    "font-awesome": "^4.7.0",
    "form-data": "^2.3.2",
    "glob-all": "^3.1.0",
    "hammerjs": "^2.0.8",
    "jquery": "3.2.1",
    "lodash-es": "^4.17.4",
    "node-object-hash": "^1.4.1",
    "popper.js": "^1.14.1",
    "redux": "^3.7.2",
    "request": "^2.88.0",
    "request-promise": "^4.2.2",
    "requestanimationframe": "0.0.23",
    "rxjs": "^6.3.3",
    "slim-select": "^1.17.0",
    "sweetalert2": "^7.26.28",
    "tooltip.js": "^1.1.7",
    "typy": "^2.0.1",
    "underscore": "~1.8",
    "whatwg-fetch": "^2.0.3"
  }

**Comment by cmatzenbach** [comment]
So I updated aurelia-framework to 1.3.0, and got a different error message this time - `TypeError: Cannot read property 'attrToRemove' of undefined` from the file aurelia-templating.js on line 3075.  Then, seeing this was in aurelia-templating, I updated all my aurelia-templating-* packages.  Each time I updated one, I got an extra version of this warning message on my console: 
`WARN [array-observation] Detected 2nd attempt of patching array from Aurelia binding. This is probably caused by dependency mismatch between core modules and a 3rd party plugin. Please see https://github.com/aurelia/cli/pull/906 if you are using webpack.`
Finally, after updating all three, I still have the same `Cannot read property 'attrToRemove' of undefined`.  I tried updating aurelia-router too, since the bug was coming from there, but it still produced the same thing.

So now I have even more questions:
1. Why did updating the aurelia-templating plugins keep adding additional binding dispatchers?
2. How can I fix this new issue?  I've debugged the problem file, and it loops through that `for` a couple times, then `expression` is eventually undefined which throws the error.

**Comment by cmatzenbach** [comment]
I kept updating the aurelia plugins, and updating aurelia-validation got me to a new error: `Error: BindingLanguage must implement inspectTextContent()` (one step further than the last error I encountered).  I updated every single aurelia package I had aside from aurelia-webpack-plugin, and this error still persisted (it's worth noting at this point I have dropped the @connectTo completely from my code and am still receiving these errors - my app won't even load).

I finally begrudgingly updated aurelia-webpack-plugin (at this point I have the error above + a total of 6 warnings of "Detected 2nd attempt of patching array..." in the console), and upgrading it to 3.0.0 broke my app completely.  I got an error about `compile` being undefined in `compiler.hooks.compile.tap(...)` - which I assume is webpack4 syntax.  Does 3.0.0 only support 4, and break support for webpack 3?

At this point I have no idea what to do.  I don't have time to upgrade my entire application to webpack 4 right now, and it seems crazy that I would need to do that just to get this decorator to work.  Is that really what I'll need to do (update everything to do with aurelia and switch to webpack4) if I want to use this?  Should this be supported for anyone not on the newest versions of everything?

**Comment by dannyBies** [comment]
Did you encounter the first error whilst running unit tests? If so it might be related to #56

**Comment by cmatzenbach** [comment]
@dannyBies No, while running my application.  I have mocha installed but haven't even written any tests yet (whoops).

**Comment by bigopon** [comment]
@cmatzenbach I'd suggest remove `node_modules` and reinstall everything. Awhile a go some folks had that issues and solved it by upgrading `aurelia-binding` to v2 and reinstalling their `node_modules`

**Comment by zewa666** [comment]
Wow seems like you went through hell debugging the issue. Sry for the trouble. Can you share your sample at the initial state so that we can try it by ourselves? Tbh it sounds more like a bundler related issue then with the Plugin itself, so its hard to say anything without an in-depth look

**Comment by cmatzenbach** [comment]
@zewa666 No worries, thanks for your help :)

I setup my initialState just as the example said to:

`export interface State {
    testProp: string;
};

export const initialState: State = {
    testProp: 'hello aurelia store!'
};`

**Comment by bigopon** [comment]
@cmatzenbach can you paste the stack trace when the exception gets thrown?

**Comment by cmatzenbach** [comment]
@bigopon So something interesting happened...I pushed my branch with the updated aurelia modules to bitbucket, created an entirely new folder and cloned the project there, installed and booted, and it started fine without errors!  However the file using aurelia-store was still using manual bindings.  When I switched to @connectTo, I got the exact same error as I did before all the module updates.  Here's the stack trace, as helpful as it is with all the webpack bundle nonsense... (which is why I debugged and broke out where the errors were occurring):

TypeError: Cannot read property 'get' of undefined
    at Object.connectTo (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:54837:106)
    at http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:165342:25
    at Object.components/findings/findings (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:165346:2)
    at __webpack_require__ (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:50:30)
    at WebpackLoader.<anonymous> (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55569:51)
    at step (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55424:23)
    at Object.next (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55405:53)
    at http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55399:71
    at 1400.__awaiter (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55395:12)
    at WebpackLoader.1400.WebpackLoader._import (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55545:16)
    at WebpackLoader.<anonymous> (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55645:44)
    at step (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55424:23)
    at Object.next (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55405:53)
    at http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55399:71
From previous event:
    at 1400.__awaiter (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55395:12)
    at WebpackLoader.1400.WebpackLoader.loadModule (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55632:16)
    at http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55623:65
    at Array.map (<anonymous>)
    at WebpackLoader.1400.WebpackLoader.loadAllModules (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55623:32)
    at ViewEngine.importViewResources (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:73503:24)
    at http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:158994:23
From previous event:
    at loadResources (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:158985:7)
    at http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:159079:14
    at next (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:158925:30)
    at runTasks (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:158931:10)
    at http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:159284:16
From previous event:
    at http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:159283:21
From previous event:
    at FrameworkConfiguration.apply (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:159267:42)
    at Aurelia.start (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:158811:37)
    at Object.<anonymous> (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:168861:50)
    at step (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:168830:23)
    at Object.next (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:168811:53)
    at http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:168805:71
From previous event:
    at main.__awaiter (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:168801:12)
    at Object.configure (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:168848:12)
    at http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55850:27
From previous event:
    at config (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55845:54)
    at http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55881:12
From previous event:
    at bootstrap (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55880:24)
    at http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55867:7
From previous event:
    at run (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55862:59)
    at Object.<anonymous> (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55887:16)
    at Object.1401 (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:55888:30)
    at __webpack_require__ (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:50:30)
    at Object.951 (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:152716:18)
    at __webpack_require__ (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:50:30)
    at 1.module.exports (http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:145:18)
    at http://cr.rsna.org:9091/app.d40849074f6e487b48bd.bundle.js:148:10

**Comment by bigopon** [comment]
@cmatzenbach What does your startup file look like? My last guess is I think you have order issue, the stacktrace is not helpful, and please put code in code block

**Comment by cmatzenbach** [comment]
Here is my main.ts, where the plugins are registered.  Is that what you mean?

```
/// <reference types="aurelia-loader-webpack/src/webpack-hot-interface"/>

// node_modules
import 'bootstrap';
import 'requestanimationframe';
import { Aurelia } from 'aurelia-framework';
import { PLATFORM } from 'aurelia-pal';
import * as Bluebird from 'bluebird';

// src
import { initialState } from 'initialState';
import 'caserepository.css';
import 'Case-Repository-Overrides.css';

// remove out if you don't want a Promise polyfill (remove also from webpack.config.js)
Bluebird.config({ warnings: false });

export async function configure(aurelia: Aurelia) {
    aurelia.use
        .standardConfiguration()
        .developmentLogging()
        .feature(PLATFORM.moduleName('components/index'))
        .feature(PLATFORM.moduleName('value-converters/index'))
        .plugin(PLATFORM.moduleName('aurelia-dialog'))
        .plugin(PLATFORM.moduleName('au-table'))
        .plugin(PLATFORM.moduleName('aurelia-validation'))
        .plugin(PLATFORM.moduleName('aurelia-store'), { initialState });
    await aurelia.start();
    await aurelia.setRoot(PLATFORM.moduleName('app'));
}
```

**Comment by bigopon** [comment]
The exception means somehow `BindingLanguage` is invoked before `aurelia.start()` happens, this is the result of a `ViewCompiler.compile` being called. From your startup, I'm unable to figure out why. I suspect it's the `components/index`. If you want, you can share screen with me and I'll be able to help. Other than that, I don't really have much input.

**Comment by cmatzenbach** [comment]
I can post my index file below in case it helps, it's just a bunch of components I'm creating that I register to be able to use globally in views.

If that doesn't help, I'd be more than happy to take you up on that offer, if you have the time.  I really appreciate it, that's really generous.

```
// node_modules
import { FrameworkConfiguration } from 'aurelia-framework';
import { PLATFORM } from 'aurelia-pal';

export function configure(config: FrameworkConfiguration) {
    config.globalResources([
        PLATFORM.moduleName('./author/author'),
        PLATFORM.moduleName('./login/login'),
        PLATFORM.moduleName('./navbar/navbar'),
        PLATFORM.moduleName('./create-case/create-case'),
        PLATFORM.moduleName('./custom-validator/custom-validator'),
        PLATFORM.moduleName('./image-upload/image-upload'),
        PLATFORM.moduleName('./case-info/case-info'),
        PLATFORM.moduleName('./subspecialty-selection/subspecialty-selection'),
        PLATFORM.moduleName('./anatomy-selection/anatomy-selection'),
        PLATFORM.moduleName('./image-tagging/image-tagging'),
        PLATFORM.moduleName('./disease-category/disease-category'),
        PLATFORM.moduleName('./exam/exam'),
        PLATFORM.moduleName('./loading-spinner/loading-spinner'),
        PLATFORM.moduleName('./image-viewer/image-viewer'),
        PLATFORM.moduleName('./image-carousel/image-carousel'),
        PLATFORM.moduleName('./image-grouping/image-grouping'),
        PLATFORM.moduleName('./image-order/image-order'),
        PLATFORM.moduleName('./references/references'),
        PLATFORM.moduleName('./diagnosis/diagnosis'),
        PLATFORM.moduleName('./pathology-images/pathology-images'),
        PLATFORM.moduleName('./home/home'),
        PLATFORM.moduleName('./navbar-public/navbar-public'),
        PLATFORM.moduleName('./case-carousel/case-carousel'),
        PLATFORM.moduleName('./view-case/view-case'),
        PLATFORM.moduleName('./coauthors/coauthors'),
        PLATFORM.moduleName('./similar-cases/similar-cases'),
        PLATFORM.moduleName('./spellcheck/spellcheck'),
        PLATFORM.moduleName('./dialog-wrapper/dialog-wrapper'),
        PLATFORM.moduleName('./findings/findings'),
    ]);
}
```

**Comment by bigopon** [comment]
@cmatzenbach lets chat on gitter then

**Comment by cmatzenbach** [comment]
Ok, the issue was fixed thanks to @bigopon (and I believe @EisenbergEffect is who he mentioned he got the idea from, correct me if it was someone else)!!!  So after all the updates (sans aurelia-webpack-loader, as we are still on webpack3 and have no desire to re-write that mess of a file to upgrade to 4), the solution ended up being deleting my node_modules folder (which I had tried numerous times before to no avail) *and* deleting the package-lock.json file.  Both of those needed to be deleted, and then re-generated on an `npm install`, and then the errors went away.

Serious thanks to @bigopon who spent at least 2-3 hours with with a complete stranger trying to figure his issue out.  All of the aurelia team members are fantastic and really care about the community.  The store has been working amazing so far and it's soooo nice to drop all the ridiculous boilerplate react brings along.  You guys did state mgmt the right way.  Thanks again!
